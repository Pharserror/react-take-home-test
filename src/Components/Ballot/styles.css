.submit-ballot-button {
  background-color: #009B86;
}

.submit-ballot-button:hover {
  background-color: #34AC9C;
}
