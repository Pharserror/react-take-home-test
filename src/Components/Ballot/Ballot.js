import './styles.css';
import Button from 'react-bootstrap/Button';
import Col from 'react-bootstrap/Col';
import React, { useCallback, useEffect, useState } from 'react';
import Row from 'react-bootstrap/Row';
import without from 'lodash/without';
import api from '../../Api/Api';
import Categories from '../Categories/Categories';
import ResultsModal from '../ResultsModal/ResultsModal';

const initialState = {
  categories: [],
  isAppHydrated: false,
  selections: {},
};

const PAGE_TITLE = 'AWARDS 2021';
export const SUBMIT_BALLOT_BUTTON_TEXT = 'SUBMIT BALLOT BUTTON';

export default function Ballot() {
  const [state, setState] = useState(initialState);
  const handleModalClose = useCallback(() => {
    setState({ ...state, areResultsDisplayed: false });
  }, [setState, state]);

  const handleNomineeSelection = useCallback((event, categoryId, nomineeId) => {
    setState({
      ...state,
      selections: {
        ...state.selections,
        [categoryId]: nomineeId,
      },
    });
  }, [setState, state]);

  const submitNomineeSelection = useCallback(() => {
    if (!state.areResultDisplayed) {
      setState({
        ...state,
        areResultsDisplayed: true,
      });
    }
  }, [setState, state])

  useEffect(async () => {
    if (!state.isAppHydrated) {
      const { items = [] } = await api.getBallotData();

      setState({ ...state, categories: items, isAppHydrated: true });
    }
  });

  return (
    <Row>
      <Col>
        <ResultsModal
          categories={state.categories}
          onHide={handleModalClose}
          selections={state.selections}
          show={state.areResultsDisplayed}
        />
        <div className="ballot">
          <Row>
            <Col>
              <h1 className="pt-5">{PAGE_TITLE}</h1>
            </Col>
          </Row>
          <Row>
            <Col>
              <Categories
                categories={state.categories}
                onClick={handleNomineeSelection}
                selections={state.selections}
              />
            </Col>
          </Row>
          <Row>
            <Col>
              <Button
                className="mt-2 mb-3 submit-ballot-button"
                onClick={submitNomineeSelection}
                size="lg"
                variant="primary"
              >
                {SUBMIT_BALLOT_BUTTON_TEXT}
              </Button>
            </Col>
          </Row>
        </div>
      </Col>
    </Row>
  );
}
