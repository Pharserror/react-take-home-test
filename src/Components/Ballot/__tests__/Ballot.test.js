import { act } from 'react-dom/test-utils';
import { render, fireEvent } from '@testing-library/react';
import Ballot, { SUBMIT_BALLOT_BUTTON_TEXT } from '../Ballot';

jest.mock('../../../Api/Api');

const defaultProps = {};

function renderComponent(props = {}) {
  return render(
    <Ballot {...props} />
  );
}

describe('default behavior', () => {
  let props;
  let renderer;

  beforeEach(() => {
    props = { ...defaultProps };
    renderer = renderComponent(props);
  });

  test('should match the snapshot', () => {
    expect(renderer).toMatchSnapshot();
  });

  describe('when I select a candidate', () => {
    beforeEach(() => {
      act(() => {
        fireEvent.click(renderer.getByText('Select Button'));
      });
    });

    describe('and I click the submit button', () => {
      beforeEach(() => {
        act(() => {
          fireEvent.click(renderer.getByText(SUBMIT_BALLOT_BUTTON_TEXT));
        });
      });

      test('I see a modal with the winner', () => {
        expect(renderer.getByText('SUCCESS')).toBeInTheDocument();
      });

      test('should match the snapshot', () => {
        expect(renderer).toMatchSnapshot();
      });
    });
  });
});
