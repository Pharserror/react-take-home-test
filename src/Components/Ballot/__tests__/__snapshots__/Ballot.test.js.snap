// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default behavior should match the snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="row"
      >
        <div
          class="col"
        >
          <div
            class="ballot"
          >
            <div
              class="row"
            >
              <div
                class="col"
              >
                <h1
                  class="pt-5"
                >
                  AWARDS 2021
                </h1>
              </div>
            </div>
            <div
              class="row"
            >
              <div
                class="col"
              >
                <div
                  class="row"
                >
                  <div
                    class="col"
                  >
                    <div
                      class="row"
                    >
                      <div
                        class="col"
                      >
                        <div
                          class="row"
                        >
                          <div
                            class="text-left col"
                          >
                            <h1
                              class="category-header"
                            >
                              Best Picture
                            </h1>
                          </div>
                        </div>
                        <div
                          class="row"
                        >
                          <div
                            class="col"
                          >
                            <div
                              class="row"
                            >
                              <div
                                class="col-lg-4 col-md-6 col-12"
                              >
                                <div
                                  class="nominee-card card"
                                >
                                  <div
                                    class="card-body"
                                  >
                                    <div
                                      class="card-title h5"
                                    >
                                      Nomadland
                                    </div>
                                    <img
                                      class="rounded-circle img-thumbnail"
                                      src="https://variety.com/wp-content/uploads/2020/12/nomadland_ver2.jpg"
                                    />
                                    <button
                                      class="btn btn-primary btn-lg"
                                      type="button"
                                    >
                                      Select Button
                                    </button>
                                  </div>
                                </div>
                              </div>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="row"
            >
              <div
                class="col"
              >
                <button
                  class="mt-2 mb-3 submit-ballot-button btn btn-primary btn-lg"
                  type="button"
                >
                  SUBMIT BALLOT BUTTON
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="row"
    >
      <div
        class="col"
      >
        <div
          class="ballot"
        >
          <div
            class="row"
          >
            <div
              class="col"
            >
              <h1
                class="pt-5"
              >
                AWARDS 2021
              </h1>
            </div>
          </div>
          <div
            class="row"
          >
            <div
              class="col"
            >
              <div
                class="row"
              >
                <div
                  class="col"
                >
                  <div
                    class="row"
                  >
                    <div
                      class="col"
                    >
                      <div
                        class="row"
                      >
                        <div
                          class="text-left col"
                        >
                          <h1
                            class="category-header"
                          >
                            Best Picture
                          </h1>
                        </div>
                      </div>
                      <div
                        class="row"
                      >
                        <div
                          class="col"
                        >
                          <div
                            class="row"
                          >
                            <div
                              class="col-lg-4 col-md-6 col-12"
                            >
                              <div
                                class="nominee-card card"
                              >
                                <div
                                  class="card-body"
                                >
                                  <div
                                    class="card-title h5"
                                  >
                                    Nomadland
                                  </div>
                                  <img
                                    class="rounded-circle img-thumbnail"
                                    src="https://variety.com/wp-content/uploads/2020/12/nomadland_ver2.jpg"
                                  />
                                  <button
                                    class="btn btn-primary btn-lg"
                                    type="button"
                                  >
                                    Select Button
                                  </button>
                                </div>
                              </div>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="row"
          >
            <div
              class="col"
            >
              <button
                class="mt-2 mb-3 submit-ballot-button btn btn-primary btn-lg"
                type="button"
              >
                SUBMIT BALLOT BUTTON
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`default behavior when I select a candidate and I click the submit button should match the snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body
    class="modal-open"
    data-rr-ui-modal-open=""
    style="padding-right: 2048px;"
  >
    <div>
      <div
        class="row"
      >
        <div
          class="col"
        >
          <div
            class="ballot"
          >
            <div
              class="row"
            >
              <div
                class="col"
              >
                <h1
                  class="pt-5"
                >
                  AWARDS 2021
                </h1>
              </div>
            </div>
            <div
              class="row"
            >
              <div
                class="col"
              >
                <div
                  class="row"
                >
                  <div
                    class="col"
                  >
                    <div
                      class="row"
                    >
                      <div
                        class="col"
                      >
                        <div
                          class="row"
                        >
                          <div
                            class="text-left col"
                          >
                            <h1
                              class="category-header"
                            >
                              Best Picture
                            </h1>
                          </div>
                        </div>
                        <div
                          class="row"
                        >
                          <div
                            class="col"
                          >
                            <div
                              class="row"
                            >
                              <div
                                class="col-lg-4 col-md-6 col-12"
                              >
                                <div
                                  class="nominee-card selected card"
                                >
                                  <div
                                    class="card-body"
                                  >
                                    <div
                                      class="card-title h5"
                                    >
                                      Nomadland
                                    </div>
                                    <img
                                      class="rounded-circle img-thumbnail"
                                      src="https://variety.com/wp-content/uploads/2020/12/nomadland_ver2.jpg"
                                    />
                                    <button
                                      class="btn btn-primary btn-lg"
                                      type="button"
                                    >
                                      Select Button
                                    </button>
                                  </div>
                                </div>
                              </div>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="row"
            >
              <div
                class="col"
              >
                <button
                  class="mt-2 mb-3 submit-ballot-button btn btn-primary btn-lg"
                  type="button"
                >
                  SUBMIT BALLOT BUTTON
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="fade modal-backdrop show"
    />
    <div
      aria-modal="true"
      class="fade modal show"
      role="dialog"
      style="display: block; padding-right: 0px;"
      tabindex="-1"
    >
      <div
        class="modal-dialog"
      >
        <div
          class="modal-content"
        >
          <div
            class="modal-header"
          >
            <div
              class="modal-title h4"
            >
              SUCCESS
            </div>
            <button
              aria-label="Close"
              class="btn-close"
              type="button"
            />
          </div>
          <div
            class="modal-body"
          >
            <div
              class="row"
            >
              <div
                class="col"
              >
                <h5>
                  The winner of 
                  Best Picture
                   is: 
                  Nomadland
                  !
                </h5>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="row"
    >
      <div
        class="col"
      >
        <div
          class="ballot"
        >
          <div
            class="row"
          >
            <div
              class="col"
            >
              <h1
                class="pt-5"
              >
                AWARDS 2021
              </h1>
            </div>
          </div>
          <div
            class="row"
          >
            <div
              class="col"
            >
              <div
                class="row"
              >
                <div
                  class="col"
                >
                  <div
                    class="row"
                  >
                    <div
                      class="col"
                    >
                      <div
                        class="row"
                      >
                        <div
                          class="text-left col"
                        >
                          <h1
                            class="category-header"
                          >
                            Best Picture
                          </h1>
                        </div>
                      </div>
                      <div
                        class="row"
                      >
                        <div
                          class="col"
                        >
                          <div
                            class="row"
                          >
                            <div
                              class="col-lg-4 col-md-6 col-12"
                            >
                              <div
                                class="nominee-card selected card"
                              >
                                <div
                                  class="card-body"
                                >
                                  <div
                                    class="card-title h5"
                                  >
                                    Nomadland
                                  </div>
                                  <img
                                    class="rounded-circle img-thumbnail"
                                    src="https://variety.com/wp-content/uploads/2020/12/nomadland_ver2.jpg"
                                  />
                                  <button
                                    class="btn btn-primary btn-lg"
                                    type="button"
                                  >
                                    Select Button
                                  </button>
                                </div>
                              </div>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="row"
          >
            <div
              class="col"
            >
              <button
                class="mt-2 mb-3 submit-ballot-button btn btn-primary btn-lg"
                type="button"
              >
                SUBMIT BALLOT BUTTON
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
